core/src/main/java/io/questdb/cutlass/line/http/LineHttpSender.java

```diff
@@ -0,0 +1,637 @@
+        }
+    }
+
+    private void escapeQuotedString(CharSequence name) {
+        for (int i = 0, n = name.length(); i < n; i++) {
+            char c = name.charAt(i);
+            switch (c) {
+                case ' ':
+                case ',':
+                case '=':
+                case '\n':
+                case '\r':
+                case '\\':
+                    request.put((byte) '\\').put((byte) c);
+                    break;
+                default:
+                    request.put(c);
+                    break;
+            }
+        }
+    }
+
+    private void escapeString(CharSequence value) {
+        for (int i = 0, n = value.length(); i < n; i++) {
+            char c = value.charAt(i);
+            switch (c) {
+                case '\n':
+                case '\r':
+                case '"':
+                case '\\':
+                    request.put((byte) '\\').put((byte) c);
+                    break;
+                default:
+                    request.put(c);
+                    break;
+            }
+        }
+    }
+
+    private void flush0(boolean closing) {
+        if (state != RequestState.EMPTY && !closing) {
+            throw new LineSenderException("Cannot flush buffer while row is in progress. Use sender.at() or sender.atNow() to finish the current row first.");
+        }
+        if (pendingRows == 0) {
+            return;
+        }
+
+        long retryingDeadlineNanos = Long.MIN_VALUE;
+        int retryBackoff = RETRY_INITIAL_BACKOFF_MS;
+        for (; ; ) {
+            try {
+                HttpClient.ResponseHeaders response = request.send(host, port);
+                response.await();
+                DirectUtf8Sequence statusCode = response.getStatusCode();
+                if (isSuccessResponse(statusCode)) {
+                    consumeChunkedResponse(response); // if any
+                    if (keepAliveDisabled(response)) {
+                        // Server has HTTP keep-alive disabled and it's closing this TCP connection.
+                        client.disconnect();
+                    }
+                    break;
+                }
+                if (isRetryableHttpStatus(statusCode)) {
+                    long nowNanos = System.nanoTime();
+                    retryingDeadlineNanos = (retryingDeadlineNanos == Long.MIN_VALUE && !closing) ? nowNanos + maxRetriesNanos : retryingDeadlineNanos;
+                    if (nowNanos >= retryingDeadlineNanos) {
+                        handleHttpErrorResponse(statusCode, response);
+                    }
+                    client.disconnect(); // forces reconnect, just in case
+                    retryBackoff = backoff(retryBackoff);
+                    continue;
+                }
+                handleHttpErrorResponse(statusCode, response);
+            } catch (HttpClientException e) {
+                // this is a network error, we can retry
```

Comment:
Why are we sure that this is a network error? If you use a non-existing host, you'll get a `HttpClientException`. While technically that's a network error (failed DNS resolve), there is not point in retrying. Probably, we should introduce type (retryable/non-retryable) to the `HttpClientException` class and check it.

Reply:
maybe an "infrastructure error" would be a better description than a "network error". 
I would prefer to do a bigger change in HTTPClient in a separate PR if that's OK with you.

Reply:
Yes, I'm fine with addressing this in a separate PR. Resolving the comment.